drivers/hmis_external_apis/app/models/hmis_external_apis/external_forms/form_generator.rb

```diff
@@ -0,0 +1,167 @@
+###
+# Copyright 2016 - 2024 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+# render html from a from definition
+module HmisExternalApis::ExternalForms
+  class FormGenerator
+    attr_reader :context
+    def initialize(context)
+      @stack = []
+      @context = context
+    end
+
+    delegate :register_field,
+             :render_form_input,
+             :render_form_textarea,
+             :render_numeric_input,
+             :render_form_date,
+             :render_form_select,
+             :render_form_radio_group,
+             :render_form_actions,
+             :render_form_fieldset,
+             :render_section,
+             :render_form_checkbox,
+             :render_dependent_block,
+             :name_from_label,
+             to: :context
+
+    def render_node(node)
+      # can turn off required fields for easier testing
+      # node['required'] = false
+      @stack.push(node)
+      result = render_dependent_item_wrapper(node) do
+        render_node_by_type(node)
+      end
+      @stack.pop
+      return result
+    end
+
+    def render_node_by_type(node)
+      node_type = node['type']
+      case node_type
+      when 'STRING'
+        render_input_node(node)
+      when 'DATE'
+        render_date_node(node)
+      when 'DISPLAY'
+        render_display_node(node)
+      when 'BOOLEAN'
+        render_boolean_node(node)
+      when 'TEXT'
+        render_text_node(node)
+      when 'GROUP'
+        render_group_node(node)
+      when 'CHOICE'
+        render_choice_node(node)
+      when 'INTEGER'
+        render_numeric_node(node)
+      else
+        raise "node type #{node_type} not supported in #{node.inspect}"
+      end
+    end
+
+    def parent_node
+      @stack[-2]
+    end
+
+    def render_input_node(node)
+      render_form_group(node: node) do
+        case node['component']
+        when 'PHONE'
+          render_form_input(label: node['text'], name: node['link_id'], required: node['required'], input_type: 'tel')
```

Comment:
I'm sure I'm missing something, and I don't think it's worth changing at this point, but it sure feels like we should have been able to lean a bit more on the rails form generation logic.  It's hard to see us re-writing something that has so many edge cases that have been handled by other tools.

Reply:
Yeah, sorry, I should have dropped this as an overall comment, not specific to that line.  I don't think we should refactor at this time, but it started to feel like `form_for` or some of the slightly lower level `input_field` and similar methods might allow us to render fields without writing custom methods and partials.  That said, this looks great, and will probably serve us for years to come.  

The edge cases I was eluding to are things like having to handle selects with multiple options turned on, or specifics around input types for integers, dates, etc.  Forms are less trouble than date handling, but a completely generic solution is probably not worth trying to implement since there are so many other form tools out there.

I agree, this implementation is reasonable and totally fine for now.
Thanks!